### PR TITLE
Upstream lemmas from Wormhole engagement

### PR DESCRIPTION
**Pull request by qian-hu**
Part of: https://github.com/runtimeverification/evm-semantics/issues/1588

This PR includes lemmas from previous Wormhole engagement.

### CODE REVIEW

File: include/kframework/word.md [hunk 1]
```diff
@@ -241,20 +241,32 @@ The maximum and minimum values of each type are defined below.
...
                  | "notMaxUInt208" [alias] /* FFFFFFFFFFFF0000000000000000000000000000000000000000000000000000 */
                  | "notMaxUInt224" [alias] /* FFFFFFFF00000000000000000000000000000000000000000000000000000000 */
+                 | "notMaxUInt240" [alias] /* FFFF000000000000000000000000000000000000000000000000000000000000 */
+                 | "notMaxUInt248" [alias] /* FF00000000000000000000000000000000000000000000000000000000000000 */
  // -------------------------------------------------------------------------------------------------------------
     rule notMaxUInt5   => 115792089237316195423570985008687907853269984665640564039457584007913129639904
+    rule notMaxUInt8 => 115792089237316195423570985008687907853269984665640564039457584007913129639680
+    rule notMaxUInt16 => 115792089237316195423570985008687907853269984665640564039457584007913129574400
+    rule notMaxUInt32 => 115792089237316195423570985008687907853269984665640564039457584007908834672640
+    rule notMaxUInt64 => 115792089237316195423570985008687907853269984665640564039439137263839420088320
```
Comment (ehildenb) [inline include/kframework/word.md:261]:
```suggestion
    rule notMaxUInt5   => 115792089237316195423570985008687907853269984665640564039457584007913129639904
    rule notMaxUInt8   => 115792089237316195423570985008687907853269984665640564039457584007913129639680
    rule notMaxUInt16  => 115792089237316195423570985008687907853269984665640564039457584007913129574400
    rule notMaxUInt32  => 115792089237316195423570985008687907853269984665640564039457584007908834672640
    rule notMaxUInt64  => 115792089237316195423570985008687907853269984665640564039439137263839420088320
```

File: include/kframework/lemmas/bitwise-simplification.k [hunk 2]
```diff
@@ -52,4 +70,16 @@ module BITWISE-SIMPLIFICATION [symbolic]
...
 
+  // ###########################################################################
+  // shift
+  // ###########################################################################
+
+    rule (X <<Int  16) <Int pow256 => true requires #rangeUInt(240, X) [simplification]
+    rule (X <<Int 240) <Int pow256 => true requires #rangeUInt( 16, X) [simplification]
+    rule (X <<Int 248) <Int pow256 => true requires #rangeUInt(  8, X) [simplification]
+
+    rule (X <<Int 16) ==Int (X *Int pow16) => true requires 0 <=Int X [simplification, smt-lemma]
```
Comment (PetarMax) [inline include/kframework/lemmas/bitwise-simplification.k:81]: I'm curious, why was this one needed?

Comment (PetarMax) [reply include/kframework/lemmas/bitwise-simplification.k:81]: ...and as an smt-lemma...

File: include/kframework/lemmas/bytes-simplification.k [hunk 3]
```diff
@@ -66,6 +66,11 @@ module BYTES-SIMPLIFICATION [symbolic]
       requires 0 <=Int W1 andBool 0 <=Int W2
       [simplification]
 
+    rule #buf(32 , #asWord(B:Bytes)) => B requires lengthBytes(B) ==Int 32 [simplification]
+
+    rule #buf(32, X <<Int 248) => #buf(1, X) +Bytes #buf(31, 0) requires #rangeUInt( 8, X) [simplification]
+    rule #buf(32, X <<Int 240) => #buf(2, X) +Bytes #buf(30, 0) requires #rangeUInt(16, X) [simplification]
```
Comment (PetarMax) [inline include/kframework/lemmas/bytes-simplification.k:72]: I think this also can be generalised like the above lemma...

**Comment by palinatolmach** [comment]
@qian-hu @PetarMax could you please provide an update on this PR? Do you think we should merge it? Thanks!

**Comment by palinatolmach** [comment]
Should be done by 10/03.

**Comment by PetarMax** [comment]
I have done a pass and added a few more lemmas - now it's up to the CI

File: include/kframework/lemmas/bytes-simplification.k (hunk 3 above)
Comment (PetarMax) [reply include/kframework/lemmas/bytes-simplification.k]: There are better lemmas for this coming from another engagement, we can merge what is here now and then add those subsequently.

**Comment by ehildenb** [comment]
I don't see any tests of the new lemmas added. Are there any from the Wormhole engagement which can be added here, like runlemma => donelemma style lemmas? We also can just write them as functional claims in some places. But we should have tests of them, regression tests are important.

**Comment by PetarMax** [comment]
> I don't see any tests of the new lemmas added. Are there any from the Wormhole engagement which can be added here, like runlemma => donelemma style lemmas? We also can just write them as functional claims in some places. But we should have tests of them, regression tests are important.

This I don't know - @qian-hu, @lucasmt?

File: tests/specs/functional/lemmas-spec.k [hunk 4]
```diff
@@ -402,6 +402,64 @@ module LEMMAS-SPEC
...
+    claim <k> runLemma ( maxUInt96  &Int ( X |Int ( notMaxUInt96  &Int Y ) ) ) => doneLemma( X ) ... </k> requires #rangeUInt (  96 , X )
+    claim <k> runLemma ( maxUInt128 &Int ( X |Int ( notMaxUInt128 &Int Y ) ) ) => doneLemma( X ) ... </k> requires #rangeUInt ( 128 , X )
+    claim <k> runLemma ( maxUInt160 &Int ( X |Int ( notMaxUInt160 &Int Y ) ) ) => doneLemma( X ) ... </k> requires #rangeUInt ( 160 , X )
+    claim <k> runLemma ( maxUInt192 &Int ( X |Int ( notMaxUInt192 &Int Y ) ) ) => doneLemma( X ) ... </k> requires #rangeUInt ( 192 , X )
+    claim <k> runLemma ( maxUInt208 &Int ( X |Int ( notMaxUInt208 &Int Y ) ) ) => doneLemma( X ) ... </k> requires #rangeUInt ( 208 , X )
+    claim <k> runLemma ( maxUInt224 &Int ( X |Int ( notMaxUInt224 &Int Y ) ) ) => doneLemma( X ) ... </k> requires #rangeUInt ( 224 , X )
+    claim <k> runLemma ( maxUInt240 &Int ( X |Int ( notMaxUInt240 &Int Y ) ) ) => doneLemma( X ) ... </k> requires #rangeUInt ( 240 , X )
+    claim <k> runLemma ( maxUInt248 &Int ( X |Int ( notMaxUInt248 &Int Y ) ) ) => doneLemma( X ) ... </k> requires #rangeUInt ( 248 , X )
```
Comment (PetarMax) [inline tests/specs/functional/lemmas-spec.k:436]: These are direct uses of the simplifications - @qian-hu, have we got anything from the engagament that needed them? I have others from the other engagement, but they require further simplifications to pass. 

I think we can merge with this for now, and then I will add mine once these further simplifications are upstreamed.

**Review by PetarMax — APPROVED** [review]

**Pull request merged**